### PR TITLE
Setup End to End Tests

### PR DESCRIPTION
**Pull request by kylepotts**
Should help start the ball following on solving https://github.com/JupiterBroadcasting/jupiterbroadcasting.com/issues/286#issuecomment-1215477732

### CODE REVIEW

**Comment by elreydetoda** [comment]
I'd like to review this one before it's merged :partying_face: 

Thank you so much @kylepotts for working on this! I'm super excited to take a look :partying_face: :rocket:

**Review by StefanS-O — APPROVED** [review]

**Comment by StefanS-O** [comment]
For me it looks good, it has it's own Github Action so it won't interfere with the build. I am not sure about storing screenshots as artifacts, because i don't know limits to that.

I have never worked with playwright, only cypress. I wonder how we could test more complicated pages with more JavaScript interaction, like the single episode page with the audio player. Looking forward to exploring that.

I think it is a good starting point for exploring further, thanks @kylepotts

**Comment by StefanS-O** [comment]
Let's see what @elreydetoda says!

**Comment by gerbrent** [comment]
nice contribution @kylepotts ! Waiting for additional feedback from our resident experts....

**Review by elreydetoda — CHANGES_REQUESTED** [review]
Overall, very nicely done @kylepotts! :partying_face: :tada: 

Everything seems to make logical sense, and I ran things locally to validate everything was passing like expected (which it was). Thank you so much for this contribution I know I personally really appreciate it and I'm sure the JB staff does too! :grin: 

Now for the review part...I know that I've left a lot of comments and I've opened [3 PRs against your repo](https://github.com/kylepotts/jupiterbroadcasting.com/pulls/elreydetoda). So, I've got a bit of feedback  :sweat_smile: , but it shouldn't be too hard because the PRs should get you most of the way (unless you want to re-implement it yourself for learning purposes).

Most of the feedback is smaller things and some best practices I've learned from videos/classes I've taken about pytest. Pretty much everything from a playwright perspective (beside the one pieces of feedback about the pause) looked really solid :grin:

Also, something that pylint has complained about me using before is that instead of `.format()` for formatting string, it normally recommends using [f-strings](https://pylint.pycqa.org/en/latest/user_guide/messages/convention/consider-using-f-string.html) instead.

Let me know if you have any questions about any of the feedback, and it might also be beneficial if you want to run [black](https://pypi.org/project/black/) & [pylint](https://pypi.org/project/pylint/) against your files as well. It makes it more cohesive, from a formatting perspective, when we're all using/linting our stuff the same way.

BTW, how was your first experience using playwright?

File: test/e2e/home.py
Comment (kylepotts) [reply test/e2e/home.py:12]: Actually I think this can probably go. I was just using it for debugging.

File: .github/workflows/e2e.yml
Comment (kylepotts) [reply .github/workflows/e2e.yml:17]: But that means that these tests don't get run on each commit which sort of defeats the purpose of having tests doesn't it? Only you plan on deploying each branch someone makes to an environment to test and then tear it down.

**Review by elreydetoda — APPROVED** [review]
After [this convo](https://matrix.to/#/!rDlEtonuSythLzmxQL:jupiterbroadcasting.com/$_S9ZPa6asy8r8dacr79RZtxcmSdcBOMwH7DhyZuJJR4?via=matrix.org&via=jupiterbroadcasting.com&via=0x6b.dev) we'll just approve this for now and I'll submit these changes later.

Just waiting on @ironicbadger's approval now.

File: .github/workflows/e2e.yml
Comment (elreydetoda) [reply .github/workflows/e2e.yml:17]: So, yes and no :upside_down_face:

My plan for deploy previews was a tiered approach.
1. Have them build on every pull request (not on JB infrastructure, but on netlify)
2. Have them run on merge into develop
3. Have them run on merge into prod

So, in that situation they would run on every push of commits, because netlify re-builds + GH actions re-run when commits are added to an already open PR.

(BTW, just to be on the same page I'm talking strictly about a future implementation)

**Review by elreydetoda — APPROVED** [review]
Awesome, it all looks good to me :grin: 

Thanks again :partying_face:

**Comment by gerbrent** [comment]
Here's my hesitation on this PR, which is worthy of much more discussion and is more a meta-topic for the project than anything and not at all a reflection of the commits/code:

While I am positive this is probably an amazing addition to our project, and really fulfills one of the main ask for the new website - i.e. "help us w your expertise to make an amazing thing together!" - it seems we are now beginning to push outward the skills of those who are in place to understand the whole project, from a longevity point of view. Speaking personally, I am just about zero help in assessing the impacts of this PR as it surpasses my skill-set. But I see, aswell, that others who are here to check on the features and inner-workings of the new website are beginning to not comprehend the entirety of what's being integrated.

@elreydetoda has clearly been an amazing asset here, so kudos Elrey!

We need to keep thinking about the 5 year datapoint: those involved in JB at a deep level in 5 years - which will necessarily be different from today - will still need to be using and understanding everything we put into this project, or at least have a roadmap of documentation/comments/explainers to help rediscover it's parts when they break (they will).

I think what I'm asking is: Can we get an explainer/documentation of what's happening here for those of us who will likely need to maintain these parts for years? I would LOVE to understand, even on a surface level, what this is solving and on simple terms, how it is solving it.

Also: ❤️ for all the hard and creative work!

**Comment by gerbrent** [comment]
Here's what I gather, from your helpful `[tests/readme.md](test/Readme.md](test/Readme.md)`, and poking around:

This sets up Playwright, a helpful testing automation framework, that will allow us to use browser engines to mimic actual user interactions, and optionally capture screenshots/screencasts of said tests. This will allow us to make sure commits/PRs do not inadvertently break unrelated features by standardizing known-true states of our working site.

[test/conftest.py](https://github.com/JupiterBroadcasting/jupiterbroadcasting.com/pull/325/files#diff-7dae78a00d8d69422a3bfbde96c7c5d794eede504a0df763853896a1ba1b5ff2) suggests expected items to look for, and presumably report on if broken. I also assume other tests can/will be written and evolved.., even for commit/branch-specific uses.

How am I doing?! ; )

File: test/Readme.md
Comment (gerbrent) [reply test/Readme.md:25]: The `Makefile` uses docker, so why not the local testing environment aswell as described in the Readme?

Comment (kylepotts) [reply test/Readme.md:25]: The Readme was made before we updated to the Dockerfile for tests, so that needs to be updated.

**Comment by kylepotts** [comment]
@gerbrent 

The goal of of testing is to make it so your code changes and deploys are inherently less risky. But nature of having the website being a software project that is going to be actively worked on we need things in place to make sure that we do not break the website with our changes. 

If you wanted to, you could deploy and manually spend time testing the site to see if you broke something. You might catch it you might not. But that manual process is painful and not very effective. 

In the world of software, there has been a lot of time and attention placed into automating these tests to sure that as we change things were not breaking them as go.

These tests are setup to run a headless browser and verify that certain elements of the page exist. They check of the header logo is still there and if the header contains the correct links in the dropdowns. It also take a screenshot of the page so someone can manually eyeball it for review to make sure nothing is out of place.

As far as your comment on the 5 year mark. These areas of software engineering are constantly evolving. There is always some new shiny testing framework. In 5 years I can guarantee these frameworks will likely be very different then they are now. Tests are not meant to be left in the dust and forgotten though. They are meant to be maintained just as you maintain the code you write. 


In the end, this was my attempt at trying to make it so there is less risk in making changes to this project that will cause problems with the jbsite when its deployed. The tests are supposed to increase the confidence that code changes will not cause issues when deployed to production. Any good software project will have tests it is in my opinion a needed and necessary part of any software project if you expect it to be maintained and worked on.  

If you decided this type of testing is not valuable because no one will know how to maintain it then that is the projects choice. I would still advocate for some kind of testing especially as you all are considering automatically deploying these changes from main into the production jb site.

**Comment by kylepotts** [comment]
@gerbrent 

I've updated the testing readme. Let me know if you would like more details on it.

**Comment by gerbrent** [comment]
@kylepotts thanks for all that!

I think the functionality is a wonderful addition to this project, just want to make sure it's something that can be used by many of us, and modified by many of us aswell. Sounds like that might not be a far stretch, so I fully support this idea given the technical bits can be confirmed/checked, which feels like @StefanS-O and @elreydetoda already confirmed.

Once you're happy w your merge cleanups, let me know @kylepotts and I'll happily merge this at a time when a few of us can be available to troubleshoot it's integration, if needed.

@kylepotts if you'd like to be involved during that merge session, we'd happily have you.

**Review by StefanS-O — APPROVED** [review]

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:6]:
```suggestion
These tests exist to make sure code changes to this repository do not break how the site currently functions. 
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:4]:
```suggestion
These tests are setup using the [Playwright](https://playwright.dev/python/docs/writing-tests) framework. They are intended to run on each commit to verify the site is working in expected order.
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:8]:
```suggestion
On every commit, these tests will be run against the committed code. The tests will verify that the site is in proper working order with respect to the changes the committer has changed.
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:10]:
```suggestion
The reason for this is so that one can feel more confident that their changes will not cause issues to the site when it is deployed in production. The goal is to catch these issues well before they make it to production thus not causing headaches for anyone maintaining the site.
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:13]:
```suggestion
These tests are written in python using the `pytest` framework in conjunction with the [Playwright](https://playwright.dev/python/docs/writing-tests) framework.
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:15]:
```suggestion
Playwright is a framework which easily allows you to spin up a headless browser and write code to verify that specific elements and content on the page exist. 
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:17]:
```suggestion
For example, you can check that the logo on the header nav exists or that a page contains a subtitle with a specific text of, for example, "Hello World"
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:19]:
```suggestion
Each file in the `e2e` should contain a test related to a certain page. The file will test for elements on that specific page only (ie. the home page, or the Coder Radio show page)
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:21]:
```suggestion
Each method in the test file will load the page it is planning to test .
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:26]:
```suggestion
And then once on that page, the test will grab elements on the page by querying the page with a selector. The selector is most commonly a css selector (or xpath) similar to what you would use with `document.querySelector`
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:30]:
```suggestion
This creates a variable called `first card` which contains the `dom` element that has a classname (ie the '.' selector) of card `<div class='.card'>`
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:32]:
```suggestion
Once there is a variable with an element, you can make assertion against the content of that element as to what you expect to be in that element.
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:37]:
```suggestion
This means we are verifying there is an element on the page that looks like this:
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:42]:
```suggestion
If such an element does exist, the test will pass. Otherwise the test will fail.
```

File: test/Readme.md
Comment (gerbrent) [inline test/Readme.md:44]:
```suggestion
You can checkout the [Playwright documentation](https://playwright.dev/docs/writing-tests) for more extensive testing methods.
```

**Review by gerbrent — APPROVED** [review]

**Comment by gerbrent** [comment]
(and someone please tell me if I'm making a bigger deal out of this merge than is needed 😄 )

**Review by gerbrent — COMMENTED** [review]
Great Readme!

**Comment by gerbrent** [comment]
This seems fun:

> Test Generator
>
> Playwright comes with the ability to generate tests out of the box and is a great way to quickly get started with testing. It will open two windows, a browser window where you interact with the website you wish to test and the Playwright Inspector window where you can record your tests, copy the tests, clear your tests as well as change the language of your tests.
> 
> Run `codegen` and perform actions in the browser. Playwright will generate the code for the user interactions. `codegen` will attempt to generate resilient text-based selectors.

https://playwright.dev/docs/codegen-intro

**Comment by gerbrent** [comment]
🎉

**Pull request merged**
